falcon/asgi/structures.py

```diff
@@ -0,0 +1,93 @@
+from json import dumps as json_dumps
+
+
+__all__ = ['SSEvent']
+
+
+class SSEvent:
+    __slots__ = [
+        'data',
+        'text',
+        'json',
+        'event',
+        'event_id',
+        'retry',
+        'comment',
+    ]
+
+    def __init__(
+        self,
+        data=None,
+        text=None,
+        json=None,
+        event=None,
+        event_id=None,
+        retry=None,
+        comment=None
+    ):
+        # NOTE(kgriffs): Check up front since this makes it a lot easier
+        #   to debug the source of the problem in the app vs. waiting for
+        #   an error to be raised from the framework when it calls serialize()
+        #   after the fact.
+
+        if data and not isinstance(data, bytes):
+            raise TypeError('data must be a byte string')
+
+        if text and not isinstance(text, str):
+            raise TypeError('text must be a string')
+
+        if event and not isinstance(event, str):
+            raise TypeError('event name must be a string')
+
+        if event_id and not isinstance(event_id, str):
+            raise TypeError('event_id must be a string')
+
+        if comment and not isinstance(comment, str):
+            raise TypeError('comment must be a string')
+
+        self.data = data
+        self.text = text
+        self.json = json
+        self.event = event
+        self.event_id = event_id
+
+        # NOTE(kgriffs) The SSE spec states that retry must be an integer;
+        #   this is a quick way to check (and coerce, if necessary) without
+        #   spending cycles on an isinstance() call.
+        self.retry = int(retry) if retry else None
```

Comment:
Just FWIW, if the value is `int`, `isinstance` seems to outperform allocation of a new `int` (CPython 3.8):
```python
>>> timeit.timeit(lambda: int(A), number=int(1e8))
8.146974843999487
>>> timeit.timeit(lambda: int(A), number=int(1e8))
8.008457058000204
>>> timeit.timeit(lambda: int(A), number=int(1e8))
8.354208095996
>>> timeit.timeit(lambda: isinstance(A, int), number=int(1e8))
6.547206504001224
>>> timeit.timeit(lambda: isinstance(A, int), number=int(1e8))
6.532925082006841
>>> timeit.timeit(lambda: isinstance(A, int), number=int(1e8))
6.530731582999579
>>> A
60
```

Reply:
I did a quick check and looks like it is a tiny bit faster on 3.6 as well. This is actually good news; isinstance used to be a lot slower! I'll make the change.